INPUT: Validation for confirm modal form

eunjuhuss opened the pull request:
#### Description:
Added validation for email confirmation code, phone confirmation code and security key.
the last 17 commits are relevant and described below.

#### Summary:
- email confirmation code
    - I have added a comment explaining the reason for email confirmation code format (uuid format) and linked the source "https://en.wikipedia.org/wiki/Universally_unique_identifier#Format"          
         - error msg(eng): Invalid confirmation code
         - error msg(swe): Ogiltig bekräftelsekod

     - max length must be 36 characters long 
           
          - error msg(eng): Email confirmation code must be 36 characters long
          - error msg(swe):  Bekräftelsekod måste vara 36 tecken


  
- phone confirmation code
    - input field must be min 10-12 characters
         - error msg(eng): Phone confirmation code must consists of 10-12 characters
         - error msg(swe): Bekräftelsekod måste innehålla 10-12 tecken

- security code
   - max length 50 characters: 
        - error msg(eng): The max length of 50 characters is reached 
        - error msg(swe): Max längden av 50 tecken har överskridits

---
- email confirmation code(valid)
<img width="1436" alt="Screenshot 2020-08-31 at 10 38 23" src="https://user-images.githubusercontent.com/44289056/91708492-b0ef4200-eb81-11ea-8f5e-9c21fd469296.png">
- email confirmation code(invalid: wrong length)
<img width="1431" alt="Screenshot 2020-08-31 at 10 38 47" src="https://user-images.githubusercontent.com/44289056/91708503-b51b5f80-eb81-11ea-868a-3bcfdc566d1b.png">
- email confirmation code(invalid: not matched uuid format)
<img width="1437" alt="Screenshot 2020-08-31 at 10 38 33" src="https://user-images.githubusercontent.com/44289056/91708501-b51b5f80-eb81-11ea-9df5-9d65502164cc.png">

- phone confirmation code(valid)
<img width="1437" alt="Screenshot 2020-08-31 at 10 39 03" src="https://user-images.githubusercontent.com/44289056/91708507-b64c8c80-eb81-11ea-98e7-70de5613b60f.png">
- phone confirmation code(invalid: wrong length)
<img width="1435" alt="Screenshot 2020-08-31 at 10 39 18" src="https://user-images.githubusercontent.com/44289056/91708505-b5b3f600-eb81-11ea-943a-e4cedaaaa18e.png">
- security code(valid)
<img width="1435" alt="Screenshot 2020-08-31 at 10 43 39" src="https://user-images.githubusercontent.com/44289056/91708496-b3ea3280-eb81-11ea-8822-a3097e71ac20.png">

- security code(invalid: over max length)
<img width="1436" alt="Screenshot 2020-08-31 at 10 43 17" src="https://user-images.githubusercontent.com/44289056/91708510-b6e52300-eb81-11ea-9847-a680ac2a795d.png">

---

#### For reviewer:
- what
- [x] Translation of error messages(Swe, Eng) are acceptable?
- [x] validations are working properly?
- [x] need more validation for confirm code?
- how
- [x] Read the above description
- [x] Reviewed the code changes
- [x] Navigate to the branch (pulled the latest version)
- [x] Run the code and been able to execute the expected function
- [ ] Check any styling on both desktop and mobile sizes

njons commented on src/login/translation/messageIndex.js:
```diff
@@ -40,12 +40,36 @@ export const unformattedMessages = defineMessages({
+  },
+  "mobile.confirm_code_wrong_length": {
+    id: "mobile.confirm_code_wrong_length",
+    defaultMessage: `Phone confirmation code must consists of 10-12 characters`,
```
extra S here: "must consist of" (but for consistency I'd consider aligning this with the 36 char message, i.e. "code must be XX characters long") and is it any character (or only letters and numbers)?

eunjuhuss replied:
English: Code must be 10-12 characters long, only letters and numbers allowed
Swedish: Koden måste vara 10-12 tecken lång, endast bokstäver och nummer tillåtna

njons commented on src/login/translation/messageIndex.js:
```diff
@@ -40,12 +40,36 @@ export const unformattedMessages = defineMessages({
+  },
+  "emails.confirm_code_wrong_length": {
+    id: "emails.confirm_code_wrong_length",
+    defaultMessage: `Email confirmation code must be 36 characters long`,
```
is it any character (or only letters and numbers)?

njons reviewed (commented):
Looks really good!

But I noticed the following: 
- the `<a>` tag runs the entire length of the modal, risking that users inadvertently click to get another code by just clicking anywhere under the input 
![Screenshot 2020-08-28 at 13 53 38](https://user-images.githubusercontent.com/30963614/91731510-2cafb580-eba7-11ea-8217-b7c2dbac6081.png)

And I had some minor reflections and questions around the messages (most added to the relevant message):
- One general thing to keep in mind (and maybe solve in the future) is that validation only happens when you click outside an input 
    - You won't know that you have the wrong length on your code or that you've reached over 50 char before you try to to save it (and you'll have to randomly cut chars and try saving again until it works), which could be frustrating

eunjuhuss commented on src/login/translation/languages/en.json:
```diff
@@ -106,6 +106,8 @@
   "emails.code_invalid": "The confirmation code is invalid, please try again or request a new code",
   "emails.code_invalid_or_expired": "The confirmation code is invalid or has expired, please try again or request a new code",
   "emails.confirm_email_placeholder": "Email confirmation code",
+  "emails.invalid_code": "Invalid confirmation code",
```
I changes text

eunjuhuss commented on src/components/ConfirmModalForm.js:
```diff
@@ -1,30 +1,44 @@
+    errors[inputName] = "required";
+  }
+  // Backend use UUID format for emailconfirmcode: https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
+  const emailPattern = /^[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}/;
```
changed emailPattern -> emailUUIDFormatPattern
phonePattern -> phoneLengthPattern

eunjuhuss commented on src/login/translation/languages/en.json:
```diff
@@ -106,6 +106,8 @@
   "emails.code_invalid_or_expired": "The confirmation code is invalid or has expired, please try again or request a new code",
   "emails.confirm_email_placeholder": "Email confirmation code",
+  "emails.invalid_code": "Invalid confirmation code",
+  "emails.confirm_code_wrong_length": "Email confirmation code must be 36 characters long",
```
I have updated the error message so If the code is not matching the uuid pattern (too short or incorrect match) the following error message will be displayed "Invalid confirmation code, only letters and numbers allowed"
If the code is longer then the uuid 36 character the following error message will be displayed "This code is too long, the code to confirm your email addresses must be 36 characters long"

njons replied:
* see new comment about dangers of being overly descriptive while trying to help users
* I could't get the over 36 char check to happen/message to show - please see screenshot in new comments and check if we need it

njons commented on src/login/translation/languages/en.json:
```diff
@@ -106,6 +106,8 @@
   "emails.code_invalid_or_expired": "The confirmation code is invalid or has expired, please try again or request a new code",
   "emails.confirm_email_placeholder": "Email confirmation code",
+  "emails.invalid_code": "Invalid confirmation code, only 32 characters and 4 hyphens allowed",
+  "emails.confirm_code_wrong_length": "This code is too long, the code to confirm your email addresses must be 36 characters long",
```
I would change "_your email address**es**_" to "_your email address_" as you only confirm one address at a time.

njons commented on src/login/translation/languages/sv.json:
```diff
@@ -106,6 +106,8 @@
   "emails.code_invalid_or_expired": "Ogiltig bekräftelsekod. Var god prova igen eller begär en ny bekräftelsekod",
   "emails.confirm_email_placeholder": "Bekräftelsekod",
+  "emails.invalid_code": "Ogiltig bekräftelsekod, endast 32 tecken och 4 bindestreck tillåtna",
+  "emails.confirm_code_wrong_length": "Den här koden är för lång, koden för att bekräfta dina e-postadresser måste vara 36 tecken lång",
```
Same comment as for English here: I would make it "_bekräfta din e-postadress_" because you only confirm one at a time

njons commented on src/login/translation/languages/sv.json:
```diff
@@ -187,6 +189,10 @@
   "mobile.button_add": "Lägg till",
   "mobile.confirm_mobile_placeholder": "Bekräftelsekod",
+  "mobile.confirm_code_wrong_length": "Koden måste vara 10-12 tecken lång, endast bokstäver och nummer tillåtna",
+  "security.confirm_security_length": "Max längden av 50 tecken har överskridits",
```
- "Maxlängden på 50 tecken har överskridits"

njons commented on src/login/translation/languages/sv.json:
```diff
@@ -187,6 +189,10 @@
+  "mobile.confirm_code_wrong_length": "Koden måste vara 10-12 tecken lång, endast bokstäver och nummer tillåtna",
+  "security.confirm_security_length": "Max längden av 50 tecken har överskridits",
+  "security.confirm_security_placeholder": "Beskriv din säkerhetsnyckel",
+  "security.help_text": "max längd av 50 tecken" ,
```
- "maxlängd på 50 tecken" - but have you considered the shorter: "max 50 tecken"?
- help text not translating to Swe: ![Screenshot 2020-09-01 at 18 04 12](https://user-images.githubusercontent.com/30963614/91877039-c184e280-ec7d-11ea-9b5b-88a37c0e0ee1.png)

njons commented on src/login/translation/languages/en.json:
```diff
@@ -187,6 +189,10 @@
   "mobile.confirm_mobile_placeholder": "Phone confirmation code",
+  "mobile.confirm_code_wrong_length": "Code must be 10-12 characters long, only letters and numbers allowed",
+  "security.confirm_security_placeholder": "Describe your security key",
+  "security.help_text": "max length of 50 characters" ,
```
have we considered "max 50 characters" here?

njons reviewed (commented):
Thanks for looking at this again! 

I have some: 
- minor comments on language
- questions: about an error message for empty values and if we match against email codes longer than 36 char (I couldn't see it)
- screenshot of one translation that is not working 

And (I know this is a legacy thing), but I noticed now that for: 
* The Swe translation for "security.webauthn-describe-title": "Add a name for your security key", is "Ge ett namn till säkerhetsnyckeln" - could we align them better

eunjuhuss commented:
- Plural to singular “adresses -> adress” for  emails.confirm_code_wrong_length 
- Swedish trnanslation text “Max längden av 50 tecken har överskridits->Max längden av 50 tecken har överskridits”
- Security help text made shorter “max length of 50 characters -> max 50 characters”
- Added translate props for security help text
- Changed “Ge ett namn till säkerhetsnyckeln -> Lägg till ett namn för säkerhetsnyckeln” for security.whbautn-describe-title
- Had same error message as label proptyps so I Removed help block prototypes. I will look into it later.

please check,

njons commented:
👍
